### PR TITLE
[IoTHubClient] Add symmetric key set api

### PR DESCRIPTION
**Pull request by hihigupt**

### CODE REVIEW

File: source/include/azure_iot_hub_client.h [hunk 1]
```diff
@@ -45,6 +45,8 @@ typedef enum AzureIoTHubClientResult
     eAzureIoTHubClientSubackWaitTimeout,  /**< There was timeout while waiting for SUBACK. */
     eAzureIoTHubClientTopicNotSubscribed, /**< Topic not subscribed. */
     eAzureIoTHubClientPublishFailed,      /**< Failed to publish. */
+    eAzureIoTHubClientSubscribeFailed,    /**< Failed to subscribe. */
+    eAzureIoTHubClientUnSubscribeFailed,  /**< Failed to unsubscribe. */
```
Comment (danewalton) [inline source/include/azure_iot_hub_client.h:49]: nit: 
```suggestion
    eAzureIoTHubClientUnsubscribeFailed,  /**< Failed to unsubscribe. */
```

File: source/azure_iot_hub_client.c [hunk 2]
```diff
@@ -583,6 +650,33 @@ void AzureIoTHubClient_Deinit( AzureIoTHubClient_t * pxAzureIoTHubClient )
 }
 /*-----------------------------------------------------------*/
 
+AzureIoTHubClientResult_t AzureIoTHubClient_SetSymmetricKey( AzureIoTHubClient_t * pxAzureIoTHubClient,
+                                                             const uint8_t * pucSymmetricKey,
+                                                             uint32_t ulSymmetricKeyLength,
+                                                             AzureIoTGetHMACFunc_t xHmacFunction )
```
Comment (danewalton) [inline source/azure_iot_hub_client.c:656]: just to unify
```suggestion
                                                             AzureIoTGetHMACFunc_t xHMACFunction )
```

File: source/azure_iot_hub_client.c [hunk 3]
```diff
@@ -460,6 +461,72 @@ static AzureIoTHubClientResult_t prvWaitForSubAck( AzureIoTHubClient_t * pxAzure
...
+    if( ulBufferLeft < ulHMACBufferLength )
+    {
+        AZLogError( ( "AzureIoTHubClient token generation failed with insufficient buffer size." ) );
+        return eAzureIoTHubClientOutOfMemory;
+    }
+
+    ulBufferLeft -= ulHMACBufferLength;
+    pucHMACBuffer = pucSASBuffer + ulSasBufferLen - ulHMACBufferLength;
+
+    if( AzureIoT_Base64HMACCalculate( pxAzureIoTHubClient->_internal.xHMACFunction,
```
Comment (danewalton) [inline source/azure_iot_hub_client.c:505]: `if () != eAzureSuccess`

File: source/azure_iot_hub_client.c [hunk 4]
```diff
@@ -460,6 +461,72 @@ static AzureIoTHubClientResult_t prvWaitForSubAck( AzureIoTHubClient_t * pxAzure
...
+                                      ucKey, ulKeyLen, pucSASBuffer, ulBytesUsed, pucSASBuffer + ulBytesUsed, ulBufferLeft,
+                                      pucHMACBuffer, ulHMACBufferLength, &ulBytesUsed ) )
+    {
+        AZLogError( ( "AzureIoTHubClient failed to encoded HMAC hash" ) );
```
Comment (danewalton) [inline source/azure_iot_hub_client.c:509]:
```suggestion
        AZLogError( ( "AzureIoTHubClient failed to encode HMAC hash" ) );
```

File: source/azure_iot_hub_client.c [hunk 5]
```diff
@@ -583,6 +650,33 @@ void AzureIoTHubClient_Deinit( AzureIoTHubClient_t * pxAzureIoTHubClient )
...
+                                                             uint32_t ulSymmetricKeyLength,
+                                                             AzureIoTGetHMACFunc_t xHmacFunction )
+{
+    AzureIoTHubClientResult_t xResult;
+
+    if( ( pxAzureIoTHubClient == NULL ) ||
+        ( pucSymmetricKey == NULL ) || ( ulSymmetricKeyLength == 0 ) ||
+        ( xHmacFunction == NULL ) )
+    {
+        AZLogError( ( "AzureIoTHubClient set symmetric key failed: Invalid argument" ) );
```
Comment (danewalton) [inline source/azure_iot_hub_client.c:664]: should this just be the function name?
```suggestion
        AZLogError( ( "AzureIoTHubClient_SetSymmetricKey failed: Invalid argument" ) );
```

File: source/azure_iot_hub_client.c [hunk 6]
```diff
@@ -583,6 +650,33 @@ void AzureIoTHubClient_Deinit( AzureIoTHubClient_t * pxAzureIoTHubClient )
...
+        xResult = eAzureIoTHubClientInvalidArgument;
+    }
+    else
+    {
+        pxAzureIoTHubClient->_internal.pucSymmetricKey = pucSymmetricKey;
+        pxAzureIoTHubClient->_internal.ulSymmetricKeyLength = ulSymmetricKeyLength;
+        pxAzureIoTHubClient->_internal.pxTokenRefresh = prvIoTHubClientGetToken;
```
Comment (danewalton) [inline source/azure_iot_hub_client.c:671]: might be a dumb question but is there ever a scenario where someone wouldn't want the refresh done automatically?

**Review by danewalton — APPROVED** [review]

File: source/azure_iot_hub_client.c (hunk 6 above)
Comment (hihigupt) [reply source/azure_iot_hub_client.c:671]: cert case, where you do not need SAS

File: source/azure_iot_hub_client.c [hunk 7]
```diff
@@ -460,6 +461,72 @@ static AzureIoTHubClientResult_t prvWaitForSubAck( AzureIoTHubClient_t * pxAzure
...
+    ulBufferLeft = ulSasBufferLen - ulBytesUsed;
+
+    if( ulBufferLeft < ulHMACBufferLength )
+    {
+        AZLogError( ( "AzureIoTHubClient token generation failed with insufficient buffer size." ) );
+        return eAzureIoTHubClientOutOfMemory;
+    }
+
+    ulBufferLeft -= ulHMACBufferLength;
+    pucHMACBuffer = pucSASBuffer + ulSasBufferLen - ulHMACBufferLength;
```
Comment (jspaith) [inline source/azure_iot_hub_client.c:503]: I think a comment here would help.  Basically you're using the last 48 bytes of the app's passed buffer to store the HMAC, the first part of the buffer is the result of the `az_iot_hub_client_sas_get_signature` call, and there's a middle section of buffer between end of `az_iot_hub_client_sas_get_signature` output and before the reserved 48 bytes that you need for AzureIoT_Base64HMACCalculate`s `pucBuffer` parameter, correct?

This all looks good, but it took me a bit to figure out how it all fits together.  So a comment about "only want caller to have to pass one buffer so we use different portions of the space as follows..." by time we're at HMAC time would've made this easier.

**Review by jspaith — APPROVED** [review]

**Pull request merged**